C error with shared struct that embeds another struct

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

**V doctor:**

```
|V full version      |V 0.4.9 4225a34.495adc9
|:-------------------|:-------------------
|OS                  |macos, macOS, 15.1.1, 24B91
|Processor           |8 cpus, 64bit, little endian, Apple M3
|Memory              |0.71GB/16GB
|                    |
|V executable        |/Users/adamoates/v/v
|V last modified time|2025-01-04 20:09:49
|                    |
|V home dir          |OK, value: /Users/adamoates/v
|VMODULES            |OK, value: /Users/adamoates/.vmodules
|VTMP                |OK, value: /tmp/v_501
|Current working dir |OK, value: /Users/adamoates/Documents/shattlebip
|                    |
|Git version         |git version 2.39.5 (Apple Git-154)
|V git status        |weekly.2024.53-29-g495adc9c
|.git/config present |true
|                    |
|cc version          |Apple clang version 16.0.0 (clang-1600.0.26.4)
|gcc version         |Apple clang version 16.0.0 (clang-1600.0.26.4)
|clang version       |Apple clang version 16.0.0 (clang-1600.0.26.4)
|tcc version         |tcc version 0.9.28rc 2024-02-05 HEAD@105d70f7 (AArch64 Darwin)
|tcc git status      |thirdparty-macos-arm64 713692d4
|emcc version        |N/A
|glibc version       |N/A
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  /Users/adamoates/Documents/test.v && /Users/adamoates/Documents/test`

```v
struct Foo {
mut:
	foo string
}

struct Bar {
	Foo
}

fn main() {
	shared bar := Bar{}
	rlock bar {
		println(bar.foo)
	}
}
```

**What did you see?**

```
================== C compilation error (from cc): ==============
cc:       |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc: /tmp/v_501/test.01JGT9FTGWQJWFRHETW6DYQQCT.tmp.c:9371:16: error: no member named 'Foo' in 'struct __shared__main__Bar'
cc:  9371 |                 println(bar->Foo.val.foo);
cc:       |                         ~~~  ^
cc: 1 warning and 1 error generated.
... (the original output was 9 lines long, and was truncated to 5 lines)
================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

Expected code to compile and print the value of `bar.foo`.

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzdhMDIzYWYxNDRmNTg1YWU1Mjg1NDkiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.5aJwJ8rypWtrE0w0GWw27IzB7hpbt2Hi6NWFWi5gqVA">Huly&reg;: <b>V_0.6-21809</b></a></sub>

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -4120,7 +4120,18 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 	if is_as_cast {
 		g.write(')')
 	}
+
+	left_is_shared := node.expr_type.has_flag(.shared_f)
+	alias_to_ptr := sym.info is ast.Alias && sym.info.parent_type.is_ptr()
+	is_dereferenced := node.expr is ast.SelectorExpr && node.expr.expr_type.is_ptr()
+		&& !node.expr.typ.is_ptr() && final_sym.kind in [.interface, .sum_type]
+	left_is_ptr := field_is_opt
+		|| (((!is_dereferenced && unwrapped_expr_type.is_ptr()) || sym.kind == .chan
+		|| alias_to_ptr) && node.from_embed_types.len == 0)
+		|| (node.expr.is_as_cast() && g.inside_smartcast)
+
 	// struct embedding
+	mut has_embed := false
 	if sym.info in [ast.Struct, ast.Aggregate] {
 		if node.generic_from_embed_types.len > 0 && sym.info is ast.Struct {
 			if sym.info.embeds.len > 0 {
@@ -4130,36 +4141,36 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 						if arr_val[0] == sym.info.embeds[0] {
 							g.write_selector_expr_embed_name(node, arr_val)
 							is_find = true
+							has_embed = true
 							break
 						}
 					}
 				}
 				if !is_find {
+					has_embed = node.from_embed_types.len > 0
 					g.write_selector_expr_embed_name(node, node.from_embed_types)
 				}
 			} else {
+				has_embed = node.from_embed_types.len > 0
 				g.write_selector_expr_embed_name(node, node.from_embed_types)
 			}
 		} else if sym.info is ast.Aggregate {
 			agg_sym := g.table.sym(sym.info.types[g.aggregate_type_idx])
 			if !g.table.struct_has_field(agg_sym, field_name) {
+				has_embed = node.from_embed_types.len > 0
 				g.write_selector_expr_embed_name(node, node.from_embed_types)
 			}
 		} else {
+			has_embed = node.from_embed_types.len > 0
 			g.write_selector_expr_embed_name(node, node.from_embed_types)
 		}
 	}
-	alias_to_ptr := sym.info is ast.Alias && sym.info.parent_type.is_ptr()
-	is_dereferenced := node.expr is ast.SelectorExpr && node.expr.expr_type.is_ptr()
-		&& !node.expr.typ.is_ptr() && final_sym.kind in [.interface, .sum_type]
-	if field_is_opt || (((!is_dereferenced && unwrapped_expr_type.is_ptr())
-		|| sym.kind == .chan || alias_to_ptr) && node.from_embed_types.len == 0)
-		|| (node.expr.is_as_cast() && g.inside_smartcast) {
+	if !has_embed && left_is_ptr {
 		g.write('->')
 	} else {
 		g.write('.')
 	}
-	if node.expr_type.has_flag(.shared_f) {
+	if !has_embed && left_is_shared {
 		g.write('val.')
 	}
 	if node.expr_type == 0 {
@@ -4239,13 +4250,17 @@ fn (mut g Gen) gen_closure_fn(expr_styp string, m ast.Fn, name string) {
 }
 
 fn (mut g Gen) write_selector_expr_embed_name(node ast.SelectorExpr, embed_types []ast.Type) {
+	is_shared := node.expr_type.has_flag(.shared_f)
 	for i, embed in embed_types {
 		embed_sym := g.table.sym(embed)
 		embed_name := embed_sym.embed_name()
 		is_left_ptr := if i == 0 {
-			node.expr_type.is_ptr()
+			node.expr_type.is_ptr() && !is_shared
 		} else {
 			embed_types[i - 1].is_ptr()
+		}
+		if i == 0 && is_shared {
+			g.write('->val')
 		}
 		if is_left_ptr {
 			g.write('->')

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -4120,16 +4120,6 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 	if is_as_cast {
 		g.write(')')
 	}
-
-	left_is_shared := node.expr_type.has_flag(.shared_f)
-	alias_to_ptr := sym.info is ast.Alias && sym.info.parent_type.is_ptr()
-	is_dereferenced := node.expr is ast.SelectorExpr && node.expr.expr_type.is_ptr()
-		&& !node.expr.typ.is_ptr() && final_sym.kind in [.interface, .sum_type]
-	left_is_ptr := field_is_opt
-		|| (((!is_dereferenced && unwrapped_expr_type.is_ptr()) || sym.kind == .chan
-		|| alias_to_ptr) && node.from_embed_types.len == 0)
-		|| (node.expr.is_as_cast() && g.inside_smartcast)
-
 	// struct embedding
 	mut has_embed := false
 	if sym.info in [ast.Struct, ast.Aggregate] {
@@ -4165,6 +4155,14 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 			g.write_selector_expr_embed_name(node, node.from_embed_types)
 		}
 	}
+	left_is_shared := node.expr_type.has_flag(.shared_f)
+	alias_to_ptr := sym.info is ast.Alias && sym.info.parent_type.is_ptr()
+	is_dereferenced := node.expr is ast.SelectorExpr && node.expr.expr_type.is_ptr()
+		&& !node.expr.typ.is_ptr() && final_sym.kind in [.interface, .sum_type]
+	left_is_ptr := field_is_opt
+		|| (((!is_dereferenced && unwrapped_expr_type.is_ptr()) || sym.kind == .chan
+		|| alias_to_ptr) && node.from_embed_types.len == 0)
+		|| (node.expr.is_as_cast() && g.inside_smartcast)
 	if !has_embed && left_is_ptr {
 		g.write('->')
 	} else {

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -4155,7 +4155,6 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 			g.write_selector_expr_embed_name(node, node.from_embed_types)
 		}
 	}
-	left_is_shared := node.expr_type.has_flag(.shared_f)
 	alias_to_ptr := sym.info is ast.Alias && sym.info.parent_type.is_ptr()
 	is_dereferenced := node.expr is ast.SelectorExpr && node.expr.expr_type.is_ptr()
 		&& !node.expr.typ.is_ptr() && final_sym.kind in [.interface, .sum_type]
@@ -4168,7 +4167,7 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 	} else {
 		g.write('.')
 	}
-	if !has_embed && left_is_shared {
+	if !has_embed && node.expr_type.has_flag(.shared_f) {
 		g.write('val.')
 	}
 	if node.expr_type == 0 {

--- a/vlib/v/tests/concurrency/selector_shared_var_test.v
+++ b/vlib/v/tests/concurrency/selector_shared_var_test.v
@@ -1,0 +1,15 @@
+struct Foo {
+mut:
+	foo string
+}
+
+struct Bar {
+	Foo
+}
+
+fn test_main() {
+	shared bar := Bar{}
+	rlock bar {
+		assert bar.foo == ''
+	}
+}

